include/history_buffer.hpp

```diff
@@ -0,0 +1,223 @@
+#ifndef GRAPH_PROTOTYPE_HISTORY_BUFFER_HPP
+#define GRAPH_PROTOTYPE_HISTORY_BUFFER_HPP
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <memory>
+#include <span>
+#include <stdexcept>
+#include <type_traits>
+#include <vector>
+
+#include <fmt/format.h>
+
+namespace gr {
+
+/**
+ * @brief A simple circular history buffer that supports contiguous ranged access across the wrap-around point.
+ *
+ * This buffer is similar to the `circular_buffer<T>` and uses a double-mapped memory approach.
+ * It is optimised for single-threaded use and does not provide thread-safety or multi-writer/multi-reader support.
```

Comment:
Will there be a way to throw an exception if multiple readers are connected to a history buffer, or make this transparent to the user?

Reply:
... not sure. Need to investigate whether this is possible w/o paying a penalty for the nominal use-case.